assembl/static2/js/app/pages/idea.jsx

```diff
@@ -9,38 +9,62 @@ import IdeaWithPosts from '../graphql/IdeaWithPosts.graphql';
 import InfiniteSeparator from '../components/common/infiniteSeparator';
 import Post, { PostFolded } from '../components/debate/thread/post';
 import GoUp from '../components/common/goUp';
-import Tree, { Child } from '../components/common/tree';
+import Tree from '../components/common/tree';
 import withLoadingIndicator from '../components/common/withLoadingIndicator';
 
 import TopPostForm from './../components/debate/thread/topPostForm';
 
-export const transformPosts = (posts) => {
+export const transformPosts = (edges, additionnalProps) => {
   const postsByParent = {};
-  posts.forEach((p) => {
+  edges.forEach((e) => {
+    const p = { ...e.node, ...additionnalProps };
     const items = postsByParent[p.parentId] || [];
     postsByParent[p.parentId] = items;
     items.push(p);
   });
 
   const getChildren = (id) => {
     return (postsByParent[id] || []).map((post) => {
-      return { ...post, children: getChildren(post.id) };
+      const newPost = post;
+      // We modify the object in place, we are sure it's already a copy from
+      // the forEach edges above.
+      newPost.children = getChildren(post.id);
+      return newPost;
     });
   };
 
   return (postsByParent.null || []).map((p) => {
-    return { ...p, children: getChildren(p.id) };
+    const newPost = p;
+    newPost.children = getChildren(p.id);
+    return newPost;
   });
 };
 
+const noRowsRenderer = () => {
+  return (
+    <div className="center">
+      <Translate value="debate.thread.noPostsInThread" />
+    </div>
+  );
+};
+
 class Idea extends React.Component {
+  componentWillMount() {
+    this.refetchIdea = this.props.data.refetch;
+  }
+
+  componentWillReceiveProps(nextProps) {
+    // data.refetchIdea is a new instance each time Idea component is rerendered
+    // so shallowEqual fails on Post components and we have a perf issue.
```

Comment:
This is indeed hackish, but I found no other way to fix this perf issue.
The issue is that the graphql() function create a component that redo the query if the props changed, even if those props are not used as variables for the query, and in the process, the data props that is passed to the wrapped component is a new instance because some methods are rebound, including refetch, see https://github.com/apollographql/react-apollo/blob/master/src/graphql.tsx#L513
so when evaluating the shouldComponentUpdate of Post components, the identity equality is false for the refetchIdea=data.refetch prop that is passed down the chain to Post components, and the posts are always rerendered.

Actually now that we removed answerPostFormId from redux store and that we only have lang in mapStateToProps, we normally can remove this hack. There should be no perf issue anymore.